run_tests.py

```diff
@@ -80,24 +83,47 @@ class GRREverythingTestLoader(test_lib.GRRTestLoader):
   base_class = test_lib.GRRBaseTest
 
 
+class GRROneTestLoader(test_lib.GRRTestLoader):
+  """Load one specific GRR test case."""
+  base_class = test_lib.GRRBaseTest
+
+  def loadTestsFromModule(self, _):
+    """Just return all the tests as if they were in the same module."""
+    test_cases = [
+        self.loadTestsFromTestCase(x) for x in self.base_class.classes.values()
+        if x.__name__ == self.test_class_name]
+
+    return self.suiteClass(test_cases)
+
+  def __init__(self, test_class_name):
+    self.test_class_name = test_class_name
+
+    super(GRROneTestLoader, self).__init__()
+
+
 def RunTest(test_suite, stream=None):
   out_fd = stream
   if stream:
     out_fd = StringIO.StringIO()
 
   try:
-    test_lib.GrrTestProgram(argv=[sys.argv[0], test_suite],
-                            testLoader=GRREverythingTestLoader(
-                                labels=flags.FLAGS.labels),
-                            testRunner=unittest.TextTestRunner(
-                                stream=out_fd))
+    program = test_lib.GrrTestProgram(
+                argv=[sys.argv[0], test_suite],
+                testLoader=GRROneTestLoader(
+                    test_class_name=test_suite),
+                testRunner=unittest.TextTestRunner(
+                    stream=out_fd))
+  except SystemExit as e:
+    # must include this section, as GrrTestProgram issues
```

Comment:
> Maybe that's the problem - they need to run in their own processes even if there is only one process. This way the test suite can be aware if a test crashes for example (otherwise it brings the entire test suite down with it).

This problem does not exist currently. I can see your point for this, but most tests fail in expected (assertion failure) or unexpected (uncaught exception) ways which are handleable by the testing framework. It doesn't make sense to require this one PR to make everything perfect right now, so I will keep it the way it is. Even if we wanted to isolate every test, running them in parallel is not a requirement; they could still run serially. I don't want to do this refactoring right now; let's file an issue if you consider it necessary. Feel free to assign it to me and I'll be happy to work on it later.

> I can see that it uses a different code path when processes == 1 - maybe we should change that.

There are many wrong things with this code. I don't want to fix everything in this PR. This PR improves the code in meaningful ways, so I suggest we merge it as it is. The rest we can report as issues on GitHub and I'll be happy to address them individually.

> The failure you are referring to shows that the test is killed probably due to OOM. We see this occasionally in Rekall too - it seems that travis does not have a lot of memory for tests and its easy to exhaust it. Maybe we should either rerun the test if its killed or consider it to pass.

I don't consider this a good option, because OOM causes the whole build to fail, not an individual test. So just marking it green in this case doesn't help us evaluate whether the tests are passing or not. OOM errors should cause the test to be red. We should also look into optimizing GRR tests for memory usage.

Rerunning is not an option. Please take a look at the build history, and you will see that the OOM error is reproducible every time on travis when multiprocessing mode is used. Therefore, I am suggesting single process mode in this PR.

> The whole idea with the test suite is that it will run all the tests as stand alone programs - this allows us to run anything including c++ tests the same way. The test is not called from a function - it is called from a runner which is forked for it.

This is not what the `run_tests.py` program did. The `run_tests.py` program had two modes of invocation: Multiprocess and single process modes. While multiprocess mode invoked single process mode, single process mode was and is invocable on its own, as you can see from the `processes` flag which can be set to 1. So, it should work for multiple tests as well. You can see that `GrrTestProgram` was expected to work in the way I am describing without exiting, because it was called within a `for` loop in single processing mode. It seems that this was never tested with multiple tests, so it is wrong, but it was [indeed called from a function](https://github.com/google/grr/blob/master/run_tests.py#L178).

As for the point of the test suite, you don't always want to run all the tests as a stand alone program each. Most testing frameworks and test runners don't do that, including Python's and nose. The behavior of running things within one process is desired for various reasons, including maintainability, simplicity, and sometimes performance depending on the context. While process-level isolation may be good practice in certain circumstances, I don't see why that's "the whole idea" of testing.

> BTW I have no idea what you mean by Google flags? We use argparse which is a standard python module.

Individual tests are not invocable on their own. The called has to go through the `run_tests.py` program, which maintains the `flags` global variable. This behavior causes problems with the `nose` test discovery mechanism. Feel free to try running `nosetests` within the grr directory to see for yourself. While this problematic behavior doesn't allow us to get rid of `run_tests.py` easily, it should be possible to replace with some effort. This should go in a different PR also.